### PR TITLE
Support top level response caching for ensemble models

### PR DESCRIPTION
**Pull request by lkomali**
ref slack thread: https://nvidia.slack.com/archives/CAZKCU4UV/p1677717244222069

Currently caching at the top-level request sent to ensemble scheduler is not supported.
Implemented caching top level requests for ensemble models. 
In case of cache hit, the response is sent without executing composing models.
In case of cache miss, the ensemble pipeline runs as is. 
Changed Logic for computing Cache Miss Latency: Cache Lookup time + Ensemble Pipeline time + Cache Insertion time
Moved similar logic form ensemble_scheduler.cc and dynamic_batch_scheduler.cc to scheduler_utils.cc to reduce code redundancy.

### CODE REVIEW

File: src/ensemble_scheduler/ensemble_scheduler/ensemble_scheduler.cc
Comment (Tabrizian) [inline src/ensemble_scheduler/ensemble_scheduler/ensemble_scheduler.cc:52]: Looks like this file is duplicate of `ensemble_scheduler.cc`

File: src/ensemble_scheduler/ensemble_scheduler.cc
Comment (rmccorm4) [inline src/ensemble_scheduler/ensemble_scheduler.cc:1497]: duplicate?

File: src/ensemble_scheduler/ensemble_scheduler.h
Comment (rmccorm4) [inline src/ensemble_scheduler/ensemble_scheduler.h:119]:
```suggestion
```

File: src/ensemble_scheduler/ensemble_scheduler.cc
Comment (rmccorm4) [inline src/ensemble_scheduler/ensemble_scheduler.cc:1448]: duplicate

File: src/ensemble_scheduler/ensemble_scheduler.cc
Comment (rmccorm4) [inline src/ensemble_scheduler/ensemble_scheduler.cc:1069]: remove

File: src/ensemble_scheduler/ensemble_scheduler.cc
Comment (rmccorm4) [inline src/ensemble_scheduler/ensemble_scheduler.cc:1071]: Do we have any tests cases that expect a failed insertion to make sure server handles it gracefully and falls back to normal inference?

File: src/ensemble_scheduler/ensemble_scheduler.cc
Comment (rmccorm4) [inline src/ensemble_scheduler/ensemble_scheduler.cc:1079]: You can remove this verbose log and join the two `ifdef TRITON_ENABLE_STATS` blocks into a single block

File: src/ensemble_scheduler/ensemble_scheduler.cc
Comment (rmccorm4) [inline src/ensemble_scheduler/ensemble_scheduler.cc:1420]: I think we can probably remove most of the verbose logging added here. Ideally the `LOG_VERBOSE(2)` at [top level](https://github.com/triton-inference-server/core/blob/56d60ccc750d5011fae084c359f4eaf3f94fb8fc/src/cache_manager.cc#L408) Insert() and Lookup() calls should be informative enough.

File: src/ensemble_scheduler/ensemble_scheduler.cc
Comment (rmccorm4) [inline src/ensemble_scheduler/ensemble_scheduler.cc:1427]: remove

File: src/ensemble_scheduler/ensemble_scheduler.cc
Comment (rmccorm4) [inline src/ensemble_scheduler/ensemble_scheduler.cc:1491]: This can be removed, as there may be multiple ensemble models, and can be queried from ensemble model config if needed

File: src/ensemble_scheduler/ensemble_scheduler.h
Comment (rmccorm4) [inline src/ensemble_scheduler/ensemble_scheduler.h:73]: Add `_` suffix for consistency with other member variables
```suggestion
  uint64_t ensemble_start_ns_;
  uint64_t ensemble_end_ns_;
```

File: src/ensemble_scheduler/ensemble_scheduler.cc
Comment (rmccorm4) [inline src/ensemble_scheduler/ensemble_scheduler.cc:1052]: Are we double counting things here if the "ensemble_end" time is captured after lookup has already finished, and then summing them together? Can you walk me through an example of the time spent on each stage, and confirm if the numbers make sense together?

File: src/ensemble_scheduler/ensemble_scheduler.cc
Comment (rmccorm4) [inline src/ensemble_scheduler/ensemble_scheduler.cc:1446]: Is this `ensemble_end_ns - ensemble_start_ns` going to include time spent in the Queue?

Comment (rmccorm4) [reply src/ensemble_scheduler/ensemble_scheduler.cc:1446]: I think sending a couple requests for the different scenarios (cache miss, cache hit), gathering all the numbers, and analyzing them to make sense would be a good step for validation here. I can help you go over the numbers after you've collected some samples and compared against your expectations.

File: src/ensemble_scheduler/ensemble_scheduler.cc
Comment (lkomali) [reply src/ensemble_scheduler/ensemble_scheduler.cc:1071]: We don't have any tests. I can add a testcase for this case.

File: src/ensemble_scheduler/ensemble_scheduler.cc
Comment (lkomali) [reply src/ensemble_scheduler/ensemble_scheduler.cc]: 1. Enqueue function calls Cache LookUp function. Time spent in lookup is captured there. If cache hit, only lookup time is stored and request is released.
2. In case of a cache hit, ensemble times are calculated just until cache insertion.
Cache insertion time is calculated separately.
I moved 
#ifdef TRITON_ENABLE_STATS
  info_->ensemble_start_ns = CaptureTimeNs();
  to line 1423 to execute just after cache lookup did not find the request.

File: src/ensemble_scheduler/ensemble_scheduler.cc
Comment (lkomali) [reply src/ensemble_scheduler/ensemble_scheduler.cc]: It doesn't capture time spent in Queue.
Request start time in the queue is captured at line 1397.
Let me know if it should capture queue time too.
Also, to avoid any confusion about time, I can remove cache lookup time and cache insert time and capture everything in ensemble_start_ns and ensemble_end_ns_.

File: src/model_repository_manager/model_lifecycle.h
Comment (rmccorm4) [inline src/model_repository_manager/model_lifecycle.h:126]: are these unrelated changes?

File: src/model_repository_manager/model_lifecycle.h
Comment (rmccorm4) [inline src/model_repository_manager/model_lifecycle.h:138]: remove

File: src/ensemble_scheduler/ensemble_scheduler.cc
Comment (rmccorm4) [reply src/ensemble_scheduler/ensemble_scheduler.cc:1485]: this one was resolved but wasn't changed

File: src/model_repository_manager/model_lifecycle.h
Comment (lkomali) [reply src/model_repository_manager/model_lifecycle.h:126]: Removed

File: src/model_repository_manager/model_lifecycle.h
Comment (lkomali) [reply src/model_repository_manager/model_lifecycle.h:138]: removed

File: src/ensemble_scheduler/ensemble_scheduler.cc
Comment (rmccorm4) [reply src/ensemble_scheduler/ensemble_scheduler.cc]: The queue **start** time is captured there, but everything from there up until the compute input starts (queue end time) is considered time spent in the queue. 

Let's chat about this on Monday, we can go over a couple examples on a whiteboard or something.

Comment (rmccorm4) [reply src/ensemble_scheduler/ensemble_scheduler.cc]: So I believe `DynamicBatchScheduler` today reports stats like this:

- Enqueue()
- request->CaptureQueueStart()
- CacheLookup() -> lookup_time = lookup_end - lookup_start
    - HIT: UpdateSuccessCacheHit()
        - queue_time = lookup_start - queue_start
        - cache_hit_time = lookup_time
    - MISS: UpdateSuccessCacheMiss()
        - (backend computation)
        - queue_time = compute_input_start - queue_start
        - (finish backend response)
        - insert_time = cache_insert_end - cache_insert_start
        - cache_miss_time =  lookup_time + insert_time (does not include compute time)

Summary:
- So in the case of cache miss, we do include some minor time spent doing lookup (should be very fast in cache miss case since there is no data copy/etc.) in the "queue time" since we do (queue_time = compute_input_start - queue_start, and compute_input_start will begin after cache_lookup_end)
- However, for cache_hit_time and cache_miss_time metrics, we only include the time spent doing actual cache things (lookup/insertion), and do not include other time spent. 
    - So I believe if we want the stats to be comparable for the ensemble and non-ensemble cases, we should do something similar
    - It looks like we [currently](https://github.com/triton-inference-server/core/blob/563a0f80a6d107c5a5fb78e46bb8bbe49538cc9a/src/ensemble_scheduler/ensemble_scheduler.cc#L1082) do: `uint64_t cache_miss_ns = lookup_ns + ensemble_ns + insert_ns;` but instead should maybe be `uint64_t cache_miss_ns = lookup_ns + insert_ns;`. What do you think?

File: src/ensemble_scheduler/ensemble_scheduler.cc
Comment (lkomali) [reply src/ensemble_scheduler/ensemble_scheduler.cc:1068]: Test added in L0_response_cache

File: src/ensemble_scheduler/ensemble_scheduler.cc
Comment (lkomali) [reply src/ensemble_scheduler/ensemble_scheduler.cc]: Fixed the calculation.
Numbers to be shared personally over email/slack.

File: src/scheduler_utils.h
Comment (oandreeva-nv) [inline src/scheduler_utils.h:37]: May I ask to provide a description of what this util is doing above its declaration?

File: src/scheduler_utils.cc
Comment (oandreeva-nv) [inline src/scheduler_utils.cc:68]: Why this needs to be in a separate block?

File: src/ensemble_scheduler/ensemble_scheduler.cc
Comment (oandreeva-nv) [inline src/ensemble_scheduler/ensemble_scheduler.cc:1433]: Do we need this line?

File: src/scheduler_utils.cc
Comment (lkomali) [reply src/scheduler_utils.cc:68]: I did not implement this function. It was written by the someone who implemented CacheLookup() function in dynamic_batcher.cc. I only moved it to a separate function in scheduler_utils.cc so that it can be reused by both dynamic scheduler and ensemble schedulers.

File: src/scheduler_utils.h
Comment (lkomali) [reply src/scheduler_utils.h:40]: Added description.

File: src/ensemble_scheduler/ensemble_scheduler.cc
Comment (lkomali) [reply src/ensemble_scheduler/ensemble_scheduler.cc]: Removed the line.

File: src/scheduler_utils.h
Comment (oandreeva-nv) [reply src/scheduler_utils.h:40]: May I ask to follow the standard we are using in the code, please. Here's the example: https://github.com/triton-inference-server/core/blob/main/src/backend_config.h

Comment (oandreeva-nv) [reply src/scheduler_utils.h:40]: Is this function is intended to be used by dynamic batch scheduler and ensemble scheduler only ?

Comment (oandreeva-nv) [reply src/scheduler_utils.h:40]: or potentially can be universal?

Comment (lkomali) [reply src/scheduler_utils.h:40]: As of now, it's being caled from these two schedulers only. I think up until this change, it's only called by dynamic batch scheduler.
But if we add more schedulers or other schedulers to support response caching, then it can be used those schedulers too.

File: src/scheduler_utils.cc
Comment (rmccorm4) [reply src/scheduler_utils.cc:68]: Probably just a logical block, no functional issues. I don't mind removing it.

Comment (oandreeva-nv) [reply src/scheduler_utils.cc:68]: Up to you and @lkomali, I was just curious if I'm missing something

**Review by rmccorm4 — APPROVED** [review]
Nice work, Harshini! 🚀

**Comment by rmccorm4** [comment]
FYI don't merge this until the testing PR is also ready to merge and pipelines look good

File: src/ensemble_scheduler/ensemble_scheduler.cc
Comment (Tabrizian) [inline src/ensemble_scheduler/ensemble_scheduler.cc:1046]: If the model is a non-decoupled model that sends the final flag and response separately is it correctly handled here?

**Review by Tabrizian — COMMENTED** [review]
Really clean code, thanks @lkomali !

File: src/ensemble_scheduler/ensemble_scheduler.cc
Comment (rmccorm4) [reply src/ensemble_scheduler/ensemble_scheduler.cc:1046]: I'm gonna file a ticket for separate investigation on this flow for both existing caching (dynamic_batch_scheduler) and this new ensemble caching if that's good with you @Tabrizian.

Comment (Tabrizian) [reply src/ensemble_scheduler/ensemble_scheduler.cc:1046]: Sounds good!

**Review by Tabrizian — APPROVED** [review]
🚀

**Pull request merged**
